### PR TITLE
feat: added support for downloading pod logs

### LINKED ISSUE #4334: Feature: Support to download logs from Dashboard
### 🔖 Feature description

This feature should enable a button to download the std-out logs that we can see in pods deployed. 

### 🎤 Pitch

This feature can be really helpful in case of air-gapped envs where all developers doesn't have the access to the platform. In such cases, there are lot of back and forth to get on a call and show all the logs and debug thing. It would make the life a lot easier if we can simply download the logs and share it with developers so that they can debug things easily. 

### ✌️ Solution

A capability to download logs.

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#7157

### PR DESCRIPTION
**Pull request by ashishdevtron**
# Description

Fixes #4334

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by kartik-579** [comment]
add semantic title

File: api/k8s/application/k8sApplicationRestHandler.go [hunk 1]
```diff
@@ -608,11 +613,99 @@ func (handler *K8sApplicationRestHandlerImpl) GetPodLogs(w http.ResponseWriter,
 		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
 		return
 	}
+	handler.requestValidationAndRBAC(w, r, token, request)
+	lastEventId := r.Header.Get("Last-Event-ID")
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRestHandler.go:617]: const

File: api/k8s/application/k8sApplicationRestHandler.go [hunk 2]
```diff
@@ -608,11 +613,99 @@ func (handler *K8sApplicationRestHandlerImpl) GetPodLogs(w http.ResponseWriter,
...
+		lastSeenMsgId, err := strconv.ParseInt(lastEventId, 10, 64)
+		if err != nil {
+			common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+			return
+		}
+		lastSeenMsgId = lastSeenMsgId + 1 //increased by one ns to avoid duplicate
+		t := v1.Unix(0, lastSeenMsgId)
+		request.K8sRequest.PodLogsRequest.SinceTime = &t
+		isReconnect = true
+	}
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRestHandler.go:629]: all numbers as const describing which value they possess

File: api/k8s/application/k8sApplicationRestHandler.go [hunk 3]
```diff
@@ -608,11 +613,99 @@ func (handler *K8sApplicationRestHandlerImpl) GetPodLogs(w http.ResponseWriter,
...
+	token := r.Header.Get("token")
+	request, err := handler.k8sApplicationService.ValidatePodLogsRequestQuery(r)
+	if err != nil {
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+	handler.requestValidationAndRBAC(w, r, token, request)
+
+	// just to make sure follow flag is set to false when downloading logs
+	request.K8sRequest.PodLogsRequest.Follow = false
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRestHandler.go:657]: in any case, can this cause nil pointer panic

File: api/k8s/application/k8sApplicationRestHandler.go [hunk 4]
```diff
@@ -608,11 +613,99 @@ func (handler *K8sApplicationRestHandlerImpl) GetPodLogs(w http.ResponseWriter,
...
+			}
+		} else if err != nil && err != io.EOF {
+			common.WriteJsonResp(w, err, nil, http.StatusInternalServerError)
+			return
+		}
+		_, err = dataBuffer.Write([]byte(log))
+		if err != nil {
+			common.WriteJsonResp(w, err, nil, http.StatusInternalServerError)
+			return
+		}
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRestHandler.go:696]: this logic is present in pump service, can extract a common method there which can be used here as well

File: api/k8s/application/k8sApplicationRestHandler.go [hunk 5]
```diff
@@ -608,11 +613,99 @@ func (handler *K8sApplicationRestHandlerImpl) GetPodLogs(w http.ResponseWriter,
...
+	}
+	podLogsFilename := fmt.Sprintf("podlogs-%s-%s.txt", request.K8sRequest.ResourceIdentifier.Name, uuid.New().String())
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRestHandler.go:698]: can move this to a small private function

File: api/k8s/application/k8sApplicationRouter.go [hunk 6]
```diff
@@ -46,9 +46,18 @@ func (impl *K8sApplicationRouterImpl) InitK8sApplicationRouter(k8sAppRouter *mux
...
 		Queries("follow", "{follow}").
-		Queries("tailLines", "{tailLines}").
+		//Queries("tailLines", "{tailLines}").
 		HandlerFunc(impl.k8sApplicationRestHandler.GetPodLogs).Methods("GET")
 
+	k8sAppRouter.Path("/pods/logs/download/{podName}").
+		Queries("containerName", "{containerName}").
+		//Queries("containerName", "{containerName}", "appId", "{appId}").
+		//Queries("clusterId", "{clusterId}", "namespace", "${namespace}").
+		//Queries("sinceSeconds", "{sinceSeconds}").
```
Comment (kartik-579) [inline api/k8s/application/k8sApplicationRouter.go:56]: remove dead code

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: api/k8s/application/k8sApplicationRestHandler.go (hunk 1 above)
Comment (ashishdevtron) [reply api/k8s/application/k8sApplicationRestHandler.go]: done

File: api/k8s/application/k8sApplicationRestHandler.go (hunk 2 above)
Comment (ashishdevtron) [reply api/k8s/application/k8sApplicationRestHandler.go]: done

File: api/k8s/application/k8sApplicationRestHandler.go (hunk 5 above)
Comment (ashishdevtron) [reply api/k8s/application/k8sApplicationRestHandler.go]: done

File: api/k8s/application/k8sApplicationRouter.go (hunk 6 above)
Comment (ashishdevtron) [reply api/k8s/application/k8sApplicationRouter.go]: done

File: api/k8s/application/k8sApplicationRestHandler.go (hunk 3 above)
Comment (ashishdevtron) [reply api/k8s/application/k8sApplicationRestHandler.go:670]: nil pointer has already been handled in the above function, which confirms that the object has already been created

**Review by vikramdevtron — APPROVED** [review]

**Review by kartik-579 — APPROVED** [review]

**Pull request merged**